### ISSUE TITLE
compiling the optimizer together with the loss

### ISSUE THREAD

**Issue by rodrigob** [lmjohns3/downhill#9]
I notice that downhill uses the CPU less efficiently than native [Keras](http://keras.io)'s optimizer. I suspect the difference is due to Keras' compiling the whole optimization loop (loss + optimizer); while downhill seems to compile the loss and the optimizer separately; and thus doing a ping-pong between the two functions.

I have not done a rigorous benchmark, but there is a noticeable difference on the CPU loads (while using the same loss and optimizer type). Have you considered using Keras' approach ?

(this would be only an internal change, downhill's API would remain the same)

**Comment by lmjohns3** [comment]
Actually downhill does compile the optimizer along with the loss. The loss is provided as a symbolic Theano expression, which gets incorporated into the update functions that downhill compiles via the symbolic gradient. This takes place in the code in the ``_compile`` method (https://github.com/lmjohns3/downhill/blob/master/downhill/base.py#L128), which compiles a Theano function to compute the monitor values given an optimization input, and then apply the updates provided by ``_get_updates``.

If you're feeling generous with your time it would be pretty great to see a profile diff between the keras and downhill SGD optimizers (to keep things simple). It might not be feasible to compare them, but hopefully there isn't *that* much difference between two SGD implementations?

**Comment by rodrigob** [comment]
My bad. I was comparing two runs with different batch sizes.
With same batch size both Keras and Downhill run in about the same time (6 minutes each for two epochs on a problem similar to `keras/examples/mnist_cnn.py`, using 40 CPU cores).

**Comment by lmjohns3** [comment]
Ok, thanks for the report! It's good to check that downhill is performing reasonably well.

I'm also curious to keep hearing about performance on a CPU-heavy machine; I only run things on a GPU machine so I don't have a good way of seeing the other hardware configurations.

**Comment by rodrigob** [comment]
Currently using CPU only because of https://github.com/fchollet/keras/issues/511
